File: java/Parser.java
import java.util.*;

class Parser {
  private Lexar lexar;
  public String parse(String input) {
    this.lexar = new Lexar(input);
    
    String expression_value = expression();
    Token token = this.lexar.getNextToken();
    
    if (token.getKind() == Token.EXPR_END) {
      return expression_value;
    } else {
      return "Error!";
    }
  }

  private String expression() {
    String component1 = number();
    List<Integer> expr_operators = Arrays.asList(Token.PLUS, Token.MINUS, Token.MULTIPLY, Token.DIVIDE);

    Token token = this.lexar.getNextToken();
    while (expr_operators.contains(token.getKind())) {
      String component2 = number();
      int c1 = Integer.parseInt(component1);
      int c2 = Integer.parseInt(component2);
      int result = 0;

      if (token.getKind() == Token.PLUS) {
        result = c1 + c2;
      } else if (token.getKind() == Token.MINUS) {
        result = c1 - c2;
      } else if (token.getKind() == Token.MULTIPLY) {
        result = c1 * c2;
      } else {
        result = c1 / c2; // Divide by zero ??
      }
      component1 = String.valueOf(result);
      token = this.lexar.getNextToken();
    };

    this.lexar.revert();
    return component1;
  }

  private String number() {
    Token token = this.lexar.getNextToken();
    String value = "";

    if (token.getKind() == Token.LPAREN) {
      value = expression();
      Token expected_rparen = this.lexar.getNextToken();
      // Throw exception if parenthesis not balanced
    } else if (token.getKind() == Token.NUMBER) {
      value = token.getValue();
    } else {
      System.out.println("Not a number");
    }
    return value;
  }
}

File: ruby/parser.rb
# require 'lexar'
require_relative 'lexar'

class Parser
  # This class takes in a string of alpha numeric
  # characters, that represent a mathematicl
  # equation. The evaluation process is simply
  # from left to right, with only given parenthesis
  # providing any order of precedence, i.e multiplcation
  # and division does not taken precedence over
  # addition or subtraction.
  #
  def parse(input)
    @lexar = Lexar.new(input)

    expression_value = expression
    token = @lexar.get_next_token
    if token.kind == Token::EXPR_END
      expression_value
    else
      raise 'End expected'
    end
  end

  protected

  def expression
    component1 = number
    expr_operators = [Token::PLUS, Token::MINUS, Token::MULTIPLY, Token::DIVIDE]

    token = @lexar.get_next_token
    while expr_operators.include?(token.kind)
      component2 = number

      if token.kind == Token::PLUS
        component1 += component2
      elsif token.kind == Token::MINUS
        component1 -= component2
      elsif token.kind == Token::MULTIPLY
        component1 *= component2
      else
        component1 /= component2
      end

      token = @lexar.get_next_token
    end

    @lexar.revert
    component1
  end

  def number
    token = @lexar.get_next_token
    
    if token.kind == Token::LPAREN
      value = expression
      expected_rparen = @lexar.get_next_token
      raise 'Unbalanced parenthesis' unless expected_rparen.kind = Token::RPAREN
    elsif token.kind == Token::NUMBER
      value = token.value
    else
      raise 'Not a number'
    end
    value
  end
end
